httpx/_urlparse.py

```diff
@@ -424,13 +424,23 @@ def is_safe(string: str, safe: str = "/") -> bool:
     """
     Determine if a given string is already quote-safe.
     """
-    NON_ESCAPED_CHARS = UNRESERVED_CHARACTERS + safe + "%"
-
-    # All characters must already be non-escaping or '%'
-    for char in string:
-        if char not in NON_ESCAPED_CHARS:
+    NON_ESCAPED_CHARS = UNRESERVED_CHARACTERS + safe
+    i = 0
+    while i < len(string):
+        char = string[i]
+        if char in NON_ESCAPED_CHARS:
+            i += 1
+            continue
+        # Check if the character is a start of a valid percent-encoded sequence
+        elif (
+            char == "%"
+            and i + 2 < len(string)
+            and string[i + 1 : i + 3].isalnum()
+            and len(string[i + 1 : i + 3]) == 2
+        ):
+            i += 3
+        else:
             return False
-
     return True
```

Comment:
```suggestion

    # All characters must already be non-escaping or valid percent-encoded
    for index, char in enumerate(string):
        if char not in NON_ESCAPED_CHARS and not PERCENT_ENCODED_REGEX.match(
            string[index : index + 3]
        ):
            return False

    return True
```

Reply:
Thanks for review! I have applied the suggestion.